BigQuery: Parsing fails if timestamps or datetimes include microsecond part

In investigating some timestamp handling, I encountered an issue that the timestamp parsing ignores microseconds if present in the timestamp. Per https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#timestamp-type the microsecond portion is optional, but it could be present.

I've updated the system tests to reflect this issue:

https://github.com/tswast/google-cloud-python/blob/5a40bf02eba937097e90bc260da2d72e16d51213/bigquery/tests/system.py#L649

This test fails with

```
$ pytest tests/system.py::TestBigQuery::test_sync_query_w_standard_sql_types
============================================================================================== test session starts ==============================================================================================
platform darwin -- Python 3.6.2, pytest-3.2.1, py-1.4.34, pluggy-0.4.0
rootdir: /Users/swast/src/google-cloud-python, inifile: setup.cfg
plugins: timeout-1.2.0
collected 1 item

tests/system.py F

=================================================================================================== FAILURES ====================================================================================================
_______________________________________________________________________________ TestBigQuery.test_sync_query_w_standard_sql_types _______________________________________________________________________________
Traceback (most recent call last):
  File "/Users/swast/src/google-cloud-python/bigquery/tests/system.py", line 747, in test_sync_query_w_standard_sql_types
    self.assertEqual(query.rows[0][0], example['expected'])
  File "/Users/swast/.pyenv/versions/3.6.2/lib/python3.6/unittest/case.py", line 829, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/Users/swast/.pyenv/versions/3.6.2/lib/python3.6/unittest/case.py", line 822, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: datetime.datetime(2016, 12, 5, 12, 41, 9, tzinfo=<UTC>) != datetime.datetime(2016, 12, 5, 12, 41, 9, 250000, tzinfo=<UTC>)
=========================================================================================== 1 failed in 5.47 seconds ============================================================================================
```

Oops, my test was incorrect. I've updated it at https://github.com/tswast/google-cloud-python/blob/bq-datetime/bigquery/tests/system.py#L680

The actual failure is

```
$ pytest tests/system.py::TestBigQuery::test_sync_query_w_standard_sql_types
============================================================= test session starts =============================================================
platform darwin -- Python 3.6.2, pytest-3.2.1, py-1.4.34, pluggy-0.4.0
rootdir: /Users/swast/src/google-cloud-python, inifile: setup.cfg
plugins: timeout-1.2.0
collected 1 item

tests/system.py F

================================================================== FAILURES ===================================================================
______________________________________________ TestBigQuery.test_sync_query_w_standard_sql_types ______________________________________________
Traceback (most recent call last):
  File "/Users/swast/src/google-cloud-python/bigquery/tests/system.py", line 746, in test_sync_query_w_standard_sql_types
    self.assertEqual(len(query.rows), 1)
  File "/Users/swast/src/google-cloud-python/bigquery/google/cloud/bigquery/query.py", line 262, in rows
    return _rows_from_json(self._properties.get('rows', ()), self.schema)
  File "/Users/swast/src/google-cloud-python/bigquery/google/cloud/bigquery/_helpers.py", line 155, in _rows_from_json
    return [_row_from_json(row, schema) for row in rows]
  File "/Users/swast/src/google-cloud-python/bigquery/google/cloud/bigquery/_helpers.py", line 155, in <listcomp>
    return [_row_from_json(row, schema) for row in rows]
  File "/Users/swast/src/google-cloud-python/bigquery/google/cloud/bigquery/_helpers.py", line 148, in _row_from_json
    row_data.append(converter(cell['v'], field))
  File "/Users/swast/src/google-cloud-python/bigquery/google/cloud/bigquery/_helpers.py", line 77, in _datetime_from_json
    return datetime.datetime.strptime(value, _RFC3339_NO_FRACTION)
  File "/Users/swast/.pyenv/versions/3.6.2/lib/python3.6/_strptime.py", line 565, in _strptime_datetime
    tt, fraction = _strptime(data_string, format)
  File "/Users/swast/.pyenv/versions/3.6.2/lib/python3.6/_strptime.py", line 365, in _strptime
    data_string[found.end():])
ValueError: unconverted data remains: .250000
========================================================== 1 failed in 8.73 seconds ===========================================================
```

--- a/bigquery/tests/system.py
+++ b/bigquery/tests/system.py
@@ -685,7 +685,8 @@ class TestBigQuery(unittest.TestCase):
                 'expected': naive,
             },
             {
-                'sql': 'SELECT DATETIME(TIMESTAMP "%s")' % (stamp_microseconds,),
+                'sql': 'SELECT DATETIME(TIMESTAMP "%s")' % (
+                    stamp_microseconds,),
                 'expected': naive_microseconds,
             },
             {

--- a/bigquery/tests/system.py
+++ b/bigquery/tests/system.py
@@ -648,6 +648,7 @@ class TestBigQuery(unittest.TestCase):
         naive = datetime.datetime(2016, 12, 5, 12, 41, 9)
         naive_microseconds = datetime.datetime(2016, 12, 5, 12, 41, 9, 250000)
         stamp = '%s %s' % (naive.date().isoformat(), naive.time().isoformat())
+        stamp_microseconds = stamp + '.250000'
         zoned = naive.replace(tzinfo=UTC)
         zoned_microseconds = naive_microseconds.replace(tzinfo=UTC)
         return [
@@ -676,7 +677,7 @@ class TestBigQuery(unittest.TestCase):
                 'expected': zoned,
             },
             {
-                'sql': 'SELECT TIMESTAMP "%s"' % (stamp,),
+                'sql': 'SELECT TIMESTAMP "%s"' % (stamp_microseconds,),
                 'expected': zoned_microseconds,
             },
             {
@@ -684,7 +685,7 @@ class TestBigQuery(unittest.TestCase):
                 'expected': naive,
             },
             {
-                'sql': 'SELECT DATETIME(TIMESTAMP "%s")' % (stamp,),
+                'sql': 'SELECT DATETIME(TIMESTAMP "%s")' % (stamp_microseconds,),
                 'expected': naive_microseconds,
             },
             {

--- a/bigquery/tests/system.py
+++ b/bigquery/tests/system.py
@@ -646,8 +646,10 @@ class TestBigQuery(unittest.TestCase):
 
     def _generate_standard_sql_types_examples(self):
         naive = datetime.datetime(2016, 12, 5, 12, 41, 9)
+        naive_microseconds = datetime.datetime(2016, 12, 5, 12, 41, 9, 250000)
         stamp = '%s %s' % (naive.date().isoformat(), naive.time().isoformat())
         zoned = naive.replace(tzinfo=UTC)
+        zoned_microseconds = naive_microseconds.replace(tzinfo=UTC)
         return [
             {
                 'sql': 'SELECT 1',
@@ -674,8 +676,16 @@ class TestBigQuery(unittest.TestCase):
                 'expected': zoned,
             },
             {
+                'sql': 'SELECT TIMESTAMP "%s"' % (stamp,),
+                'expected': zoned_microseconds,
+            },
+            {
                 'sql': 'SELECT DATETIME(TIMESTAMP "%s")' % (stamp,),
                 'expected': naive,
+            },
+            {
+                'sql': 'SELECT DATETIME(TIMESTAMP "%s")' % (stamp,),
+                'expected': naive_microseconds,
             },
             {
                 'sql': 'SELECT DATE(TIMESTAMP "%s")' % (stamp,),

